<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.6.1 -->
<title>koa 中的错误处理 | 牛さんの部落格</title>
<meta name="generator" content="Jekyll v4.1.1" />
<meta property="og:title" content="koa 中的错误处理" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="koa 中的错误处理 不像 express 中在末尾处注册一个声明为 (err, req, res, next) 中间件的方式，koa 刚好相反，在开头进行注册。 app.use(async (ctx, next) =&gt; { try { await next(); } catch (err) { ctx.status = err.status || 500; ctx.body = err.message; ctx.app.emit(&quot;error&quot;, err, ctx); } }); 这样程序中任何报错都会收敛到此处。此时可以方便地将错误打印到页面，开发时非常便捷。 + ctx.app.emit(&#39;error&#39;, err, ctx); koa 也建议通过 app 来派发错误，然后通过监听 app 上的 error 事件对这些错误做进一步的统一处理和集中管理。 app.on(&quot;error&quot;, (err, ctx) =&gt; { /* 错误的集中处理: * log 出来 * 写入日志 * 写入数据库 * ... */ }); 一个错误捕获并打印到页面的示例： const Koa = require(&quot;koa&quot;); const app = new Koa(); app.use(async (ctx, next) =&gt; { try { await next(); } catch (err) { const status = err.status || 500; ctx.status = status; ctx.type = &quot;html&quot;; ctx.body = ` &lt;b&gt;${status}&lt;/b&gt; ${err} `; // emmit ctx.app.emit(&quot;error&quot;, err, ctx); } }); app.use(ctx =&gt; { const a = &quot;hello&quot;; a = &quot;hello world!&quot;; // TypeError: Assignment to constant variable. ctx.body = a; }); app.on(&quot;error&quot;, (err, ctx) =&gt; { console.error(&quot;Ooops..\n&quot;, err); }); app.listen(3000); 通过 node server.js 启动后访问页面可看到命令行的错误输出。 如果使用 pm2，可通过 —no-daemon 参数使其停留在在命令行以查看输出。 如果不使用上述参数，可通过 pm2 logs [app-name] 来查看。 ctx.throw 朴素的抛错方式需要手动设置状态码及信息对客户端的可见性。 const err = new Error(&quot;err msg&quot;); err.status = 401; err.expose = true; throw err; expose 决定是否会返回错误详情给客户端，否则只展示状态对应的错误文案，比如 500 会在浏览器中展示为 Internal Server Error。 而通过 ctx.throw 这个 helper 方法会更加简洁。 上面的代码片段等价于: ctx.throw(401, &quot;err msg&quot;); 如果不指定状态码，默认为 500。5xx 类错误 expose 默认为 false，即不会将错误信息返回到 response。 抛错时还可以传递一些额外数据，这些数据会合并到错误对象上，在处理错误的地方可以从 error 上获取。 app.use(ctx =&gt; { ctx.throw(401, &quot;access_denied&quot;, { user: { name: &quot;foo&quot; } }); }); app.on(&quot;error&quot;, (err, ctx) =&gt; { console.error(&quot;Ooops..\n&quot;, err.user); }); 参考 Error Handling ctx.throw" />
<meta property="og:description" content="koa 中的错误处理 不像 express 中在末尾处注册一个声明为 (err, req, res, next) 中间件的方式，koa 刚好相反，在开头进行注册。 app.use(async (ctx, next) =&gt; { try { await next(); } catch (err) { ctx.status = err.status || 500; ctx.body = err.message; ctx.app.emit(&quot;error&quot;, err, ctx); } }); 这样程序中任何报错都会收敛到此处。此时可以方便地将错误打印到页面，开发时非常便捷。 + ctx.app.emit(&#39;error&#39;, err, ctx); koa 也建议通过 app 来派发错误，然后通过监听 app 上的 error 事件对这些错误做进一步的统一处理和集中管理。 app.on(&quot;error&quot;, (err, ctx) =&gt; { /* 错误的集中处理: * log 出来 * 写入日志 * 写入数据库 * ... */ }); 一个错误捕获并打印到页面的示例： const Koa = require(&quot;koa&quot;); const app = new Koa(); app.use(async (ctx, next) =&gt; { try { await next(); } catch (err) { const status = err.status || 500; ctx.status = status; ctx.type = &quot;html&quot;; ctx.body = ` &lt;b&gt;${status}&lt;/b&gt; ${err} `; // emmit ctx.app.emit(&quot;error&quot;, err, ctx); } }); app.use(ctx =&gt; { const a = &quot;hello&quot;; a = &quot;hello world!&quot;; // TypeError: Assignment to constant variable. ctx.body = a; }); app.on(&quot;error&quot;, (err, ctx) =&gt; { console.error(&quot;Ooops..\n&quot;, err); }); app.listen(3000); 通过 node server.js 启动后访问页面可看到命令行的错误输出。 如果使用 pm2，可通过 —no-daemon 参数使其停留在在命令行以查看输出。 如果不使用上述参数，可通过 pm2 logs [app-name] 来查看。 ctx.throw 朴素的抛错方式需要手动设置状态码及信息对客户端的可见性。 const err = new Error(&quot;err msg&quot;); err.status = 401; err.expose = true; throw err; expose 决定是否会返回错误详情给客户端，否则只展示状态对应的错误文案，比如 500 会在浏览器中展示为 Internal Server Error。 而通过 ctx.throw 这个 helper 方法会更加简洁。 上面的代码片段等价于: ctx.throw(401, &quot;err msg&quot;); 如果不指定状态码，默认为 500。5xx 类错误 expose 默认为 false，即不会将错误信息返回到 response。 抛错时还可以传递一些额外数据，这些数据会合并到错误对象上，在处理错误的地方可以从 error 上获取。 app.use(ctx =&gt; { ctx.throw(401, &quot;access_denied&quot;, { user: { name: &quot;foo&quot; } }); }); app.on(&quot;error&quot;, (err, ctx) =&gt; { console.error(&quot;Ooops..\n&quot;, err.user); }); 参考 Error Handling ctx.throw" />
<link rel="canonical" href="http://localhost:4000/2019/04/09/koa-%E4%B8%AD%E7%9A%84%E9%94%99%E8%AF%AF%E5%A4%84%E7%90%86.html" />
<meta property="og:url" content="http://localhost:4000/2019/04/09/koa-%E4%B8%AD%E7%9A%84%E9%94%99%E8%AF%AF%E5%A4%84%E7%90%86.html" />
<meta property="og:site_name" content="牛さんの部落格" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2019-04-09T23:04:00+08:00" />
<script type="application/ld+json">
{"url":"http://localhost:4000/2019/04/09/koa-%E4%B8%AD%E7%9A%84%E9%94%99%E8%AF%AF%E5%A4%84%E7%90%86.html","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/2019/04/09/koa-%E4%B8%AD%E7%9A%84%E9%94%99%E8%AF%AF%E5%A4%84%E7%90%86.html"},"description":"koa 中的错误处理 不像 express 中在末尾处注册一个声明为 (err, req, res, next) 中间件的方式，koa 刚好相反，在开头进行注册。 app.use(async (ctx, next) =&gt; { try { await next(); } catch (err) { ctx.status = err.status || 500; ctx.body = err.message; ctx.app.emit(&quot;error&quot;, err, ctx); } }); 这样程序中任何报错都会收敛到此处。此时可以方便地将错误打印到页面，开发时非常便捷。 + ctx.app.emit(&#39;error&#39;, err, ctx); koa 也建议通过 app 来派发错误，然后通过监听 app 上的 error 事件对这些错误做进一步的统一处理和集中管理。 app.on(&quot;error&quot;, (err, ctx) =&gt; { /* 错误的集中处理: * log 出来 * 写入日志 * 写入数据库 * ... */ }); 一个错误捕获并打印到页面的示例： const Koa = require(&quot;koa&quot;); const app = new Koa(); app.use(async (ctx, next) =&gt; { try { await next(); } catch (err) { const status = err.status || 500; ctx.status = status; ctx.type = &quot;html&quot;; ctx.body = ` &lt;b&gt;${status}&lt;/b&gt; ${err} `; // emmit ctx.app.emit(&quot;error&quot;, err, ctx); } }); app.use(ctx =&gt; { const a = &quot;hello&quot;; a = &quot;hello world!&quot;; // TypeError: Assignment to constant variable. ctx.body = a; }); app.on(&quot;error&quot;, (err, ctx) =&gt; { console.error(&quot;Ooops..\\n&quot;, err); }); app.listen(3000); 通过 node server.js 启动后访问页面可看到命令行的错误输出。 如果使用 pm2，可通过 —no-daemon 参数使其停留在在命令行以查看输出。 如果不使用上述参数，可通过 pm2 logs [app-name] 来查看。 ctx.throw 朴素的抛错方式需要手动设置状态码及信息对客户端的可见性。 const err = new Error(&quot;err msg&quot;); err.status = 401; err.expose = true; throw err; expose 决定是否会返回错误详情给客户端，否则只展示状态对应的错误文案，比如 500 会在浏览器中展示为 Internal Server Error。 而通过 ctx.throw 这个 helper 方法会更加简洁。 上面的代码片段等价于: ctx.throw(401, &quot;err msg&quot;); 如果不指定状态码，默认为 500。5xx 类错误 expose 默认为 false，即不会将错误信息返回到 response。 抛错时还可以传递一些额外数据，这些数据会合并到错误对象上，在处理错误的地方可以从 error 上获取。 app.use(ctx =&gt; { ctx.throw(401, &quot;access_denied&quot;, { user: { name: &quot;foo&quot; } }); }); app.on(&quot;error&quot;, (err, ctx) =&gt; { console.error(&quot;Ooops..\\n&quot;, err.user); }); 参考 Error Handling ctx.throw","@type":"BlogPosting","headline":"koa 中的错误处理","dateModified":"2019-04-09T23:04:00+08:00","datePublished":"2019-04-09T23:04:00+08:00","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="牛さんの部落格" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/">牛さんの部落格</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/about/">About</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">koa 中的错误处理</h1>
    <p class="post-meta">
      <time class="dt-published" datetime="2019-04-09T23:04:00+08:00" itemprop="datePublished">Apr 9, 2019
      </time></p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <h1 id="koa-中的错误处理">koa 中的错误处理</h1>

<p>不像 express 中在末尾处注册一个声明为 <code class="language-plaintext highlighter-rouge">(err, req, res, next)</code> 中间件的方式，koa 刚好相反，在开头进行注册。</p>

<div class="language-js highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nx">app</span><span class="p">.</span><span class="nx">use</span><span class="p">(</span><span class="k">async</span> <span class="p">(</span><span class="nx">ctx</span><span class="p">,</span> <span class="nx">next</span><span class="p">)</span> <span class="o">=&gt;</span> <span class="p">{</span>
  <span class="k">try</span> <span class="p">{</span>
    <span class="k">await</span> <span class="nx">next</span><span class="p">();</span>
  <span class="p">}</span> <span class="k">catch</span> <span class="p">(</span><span class="nx">err</span><span class="p">)</span> <span class="p">{</span>
    <span class="nx">ctx</span><span class="p">.</span><span class="nx">status</span> <span class="o">=</span> <span class="nx">err</span><span class="p">.</span><span class="nx">status</span> <span class="o">||</span> <span class="mi">500</span><span class="p">;</span>
    <span class="nx">ctx</span><span class="p">.</span><span class="nx">body</span> <span class="o">=</span> <span class="nx">err</span><span class="p">.</span><span class="nx">message</span><span class="p">;</span>
    <span class="nx">ctx</span><span class="p">.</span><span class="nx">app</span><span class="p">.</span><span class="nx">emit</span><span class="p">(</span><span class="dl">"</span><span class="s2">error</span><span class="dl">"</span><span class="p">,</span> <span class="nx">err</span><span class="p">,</span> <span class="nx">ctx</span><span class="p">);</span>
  <span class="p">}</span>
<span class="p">});</span>
</code></pre></div></div>

<p>这样程序中任何报错都会收敛到此处。此时可以方便地将错误打印到页面，开发时非常便捷。</p>

<div class="language-diff highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="gi">+     ctx.app.emit('error', err, ctx);
</span></code></pre></div></div>

<p>koa 也建议通过 app 来派发错误，然后通过监听 app 上的 <code class="language-plaintext highlighter-rouge">error</code> 事件对这些错误做进一步的统一处理和集中管理。</p>

<div class="language-js highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nx">app</span><span class="p">.</span><span class="nx">on</span><span class="p">(</span><span class="dl">"</span><span class="s2">error</span><span class="dl">"</span><span class="p">,</span> <span class="p">(</span><span class="nx">err</span><span class="p">,</span> <span class="nx">ctx</span><span class="p">)</span> <span class="o">=&gt;</span> <span class="p">{</span>
  <span class="cm">/* 错误的集中处理:
   *  log 出来
   *  写入日志
   *  写入数据库
   *   ...
   */</span>
<span class="p">});</span>
</code></pre></div></div>

<p>一个错误捕获并打印到页面的示例：</p>

<div class="language-js highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kd">const</span> <span class="nx">Koa</span> <span class="o">=</span> <span class="nx">require</span><span class="p">(</span><span class="dl">"</span><span class="s2">koa</span><span class="dl">"</span><span class="p">);</span>
<span class="kd">const</span> <span class="nx">app</span> <span class="o">=</span> <span class="k">new</span> <span class="nx">Koa</span><span class="p">();</span>

<span class="nx">app</span><span class="p">.</span><span class="nx">use</span><span class="p">(</span><span class="k">async</span> <span class="p">(</span><span class="nx">ctx</span><span class="p">,</span> <span class="nx">next</span><span class="p">)</span> <span class="o">=&gt;</span> <span class="p">{</span>
  <span class="k">try</span> <span class="p">{</span>
    <span class="k">await</span> <span class="nx">next</span><span class="p">();</span>
  <span class="p">}</span> <span class="k">catch</span> <span class="p">(</span><span class="nx">err</span><span class="p">)</span> <span class="p">{</span>
    <span class="kd">const</span> <span class="nx">status</span> <span class="o">=</span> <span class="nx">err</span><span class="p">.</span><span class="nx">status</span> <span class="o">||</span> <span class="mi">500</span><span class="p">;</span>
    <span class="nx">ctx</span><span class="p">.</span><span class="nx">status</span> <span class="o">=</span> <span class="nx">status</span><span class="p">;</span>
    <span class="nx">ctx</span><span class="p">.</span><span class="nx">type</span> <span class="o">=</span> <span class="dl">"</span><span class="s2">html</span><span class="dl">"</span><span class="p">;</span>
    <span class="nx">ctx</span><span class="p">.</span><span class="nx">body</span> <span class="o">=</span> <span class="s2">`
    &lt;b&gt;</span><span class="p">${</span><span class="nx">status</span><span class="p">}</span><span class="s2">&lt;/b&gt; </span><span class="p">${</span><span class="nx">err</span><span class="p">}</span><span class="s2">
    `</span><span class="p">;</span>
    <span class="c1">// emmit</span>
    <span class="nx">ctx</span><span class="p">.</span><span class="nx">app</span><span class="p">.</span><span class="nx">emit</span><span class="p">(</span><span class="dl">"</span><span class="s2">error</span><span class="dl">"</span><span class="p">,</span> <span class="nx">err</span><span class="p">,</span> <span class="nx">ctx</span><span class="p">);</span>
  <span class="p">}</span>
<span class="p">});</span>

<span class="nx">app</span><span class="p">.</span><span class="nx">use</span><span class="p">(</span><span class="nx">ctx</span> <span class="o">=&gt;</span> <span class="p">{</span>
  <span class="kd">const</span> <span class="nx">a</span> <span class="o">=</span> <span class="dl">"</span><span class="s2">hello</span><span class="dl">"</span><span class="p">;</span>
  <span class="nx">a</span> <span class="o">=</span> <span class="dl">"</span><span class="s2">hello world!</span><span class="dl">"</span><span class="p">;</span> <span class="c1">// TypeError: Assignment to constant variable.</span>
  <span class="nx">ctx</span><span class="p">.</span><span class="nx">body</span> <span class="o">=</span> <span class="nx">a</span><span class="p">;</span>
<span class="p">});</span>

<span class="nx">app</span><span class="p">.</span><span class="nx">on</span><span class="p">(</span><span class="dl">"</span><span class="s2">error</span><span class="dl">"</span><span class="p">,</span> <span class="p">(</span><span class="nx">err</span><span class="p">,</span> <span class="nx">ctx</span><span class="p">)</span> <span class="o">=&gt;</span> <span class="p">{</span>
  <span class="nx">console</span><span class="p">.</span><span class="nx">error</span><span class="p">(</span><span class="dl">"</span><span class="s2">Ooops..</span><span class="se">\n</span><span class="dl">"</span><span class="p">,</span> <span class="nx">err</span><span class="p">);</span>
<span class="p">});</span>

<span class="nx">app</span><span class="p">.</span><span class="nx">listen</span><span class="p">(</span><span class="mi">3000</span><span class="p">);</span>
</code></pre></div></div>

<p>通过 <code class="language-plaintext highlighter-rouge">node server.js</code> 启动后访问页面可看到命令行的错误输出。
如果使用 pm2，可通过 <code class="language-plaintext highlighter-rouge">—no-daemon</code> 参数使其停留在在命令行以查看输出。
如果不使用上述参数，可通过 <code class="language-plaintext highlighter-rouge">pm2 logs [app-name]</code> 来查看。</p>

<h3 id="ctxthrow"><code class="language-plaintext highlighter-rouge">ctx.throw</code></h3>

<p>朴素的抛错方式需要手动设置状态码及信息对客户端的可见性。</p>

<div class="language-js highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kd">const</span> <span class="nx">err</span> <span class="o">=</span> <span class="k">new</span> <span class="nb">Error</span><span class="p">(</span><span class="dl">"</span><span class="s2">err msg</span><span class="dl">"</span><span class="p">);</span>
<span class="nx">err</span><span class="p">.</span><span class="nx">status</span> <span class="o">=</span> <span class="mi">401</span><span class="p">;</span>
<span class="nx">err</span><span class="p">.</span><span class="nx">expose</span> <span class="o">=</span> <span class="kc">true</span><span class="p">;</span>
<span class="k">throw</span> <span class="nx">err</span><span class="p">;</span>
</code></pre></div></div>

<p><code class="language-plaintext highlighter-rouge">expose</code> 决定是否会返回错误详情给客户端，否则只展示状态对应的错误文案，比如 500 会在浏览器中展示为 <code class="language-plaintext highlighter-rouge">Internal Server Error</code>。</p>

<p>而通过 <a href="https://github.com/koajs/koa/blob/master/docs/api/context.md#ctxthrowstatus-msg-properties"><code class="language-plaintext highlighter-rouge">ctx.throw</code></a> 这个 helper 方法会更加简洁。
上面的代码片段等价于:</p>

<div class="language-js highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nx">ctx</span><span class="p">.</span><span class="k">throw</span><span class="p">(</span><span class="mi">401</span><span class="p">,</span> <span class="dl">"</span><span class="s2">err msg</span><span class="dl">"</span><span class="p">);</span>
</code></pre></div></div>

<p>如果不指定状态码，默认为 500。5xx 类错误 <code class="language-plaintext highlighter-rouge">expose</code> 默认为 <code class="language-plaintext highlighter-rouge">false</code>，即不会将错误信息返回到 response。</p>

<p>抛错时还可以传递一些额外数据，这些数据会合并到错误对象上，在处理错误的地方可以从 <code class="language-plaintext highlighter-rouge">error</code> 上获取。</p>

<div class="language-js highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nx">app</span><span class="p">.</span><span class="nx">use</span><span class="p">(</span><span class="nx">ctx</span> <span class="o">=&gt;</span> <span class="p">{</span>
  <span class="nx">ctx</span><span class="p">.</span><span class="k">throw</span><span class="p">(</span><span class="mi">401</span><span class="p">,</span> <span class="dl">"</span><span class="s2">access_denied</span><span class="dl">"</span><span class="p">,</span> <span class="p">{</span> <span class="na">user</span><span class="p">:</span> <span class="p">{</span> <span class="na">name</span><span class="p">:</span> <span class="dl">"</span><span class="s2">foo</span><span class="dl">"</span> <span class="p">}</span> <span class="p">});</span>
<span class="p">});</span>

<span class="nx">app</span><span class="p">.</span><span class="nx">on</span><span class="p">(</span><span class="dl">"</span><span class="s2">error</span><span class="dl">"</span><span class="p">,</span> <span class="p">(</span><span class="nx">err</span><span class="p">,</span> <span class="nx">ctx</span><span class="p">)</span> <span class="o">=&gt;</span> <span class="p">{</span>
  <span class="nx">console</span><span class="p">.</span><span class="nx">error</span><span class="p">(</span><span class="dl">"</span><span class="s2">Ooops..</span><span class="se">\n</span><span class="dl">"</span><span class="p">,</span> <span class="nx">err</span><span class="p">.</span><span class="nx">user</span><span class="p">);</span>
<span class="p">});</span>
</code></pre></div></div>

<h3 id="参考">参考</h3>

<ul>
  <li><a href="https://github.com/koajs/koa/wiki/Error-Handling">Error Handling</a></li>
  <li><a href="https://github.com/koajs/koa/blob/master/docs/api/context.md#ctxthrowstatus-msg-properties"><code class="language-plaintext highlighter-rouge">ctx.throw</code></a></li>
</ul>


  </div><a class="u-url" href="/2019/04/09/koa-%E4%B8%AD%E7%9A%84%E9%94%99%E8%AF%AF%E5%A4%84%E7%90%86.html" hidden></a>
</article>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">牛さんの部落格</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">牛さんの部落格</li><li><a class="u-email" href="mailto:liuwayong@gmail.com">liuwayong@gmail.com</a></li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/jekyll"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg> <span class="username">jekyll</span></a></li><li><a href="https://www.twitter.com/jekyllrb"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#twitter"></use></svg> <span class="username">jekyllrb</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>notes about the coding life</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
